beacon-chain/blockchain/service.go

```diff
@@ -120,118 +123,249 @@ func NewService(ctx context.Context, opts ...Option) (*Service, error) {
+		// Exit run time if the node failed to verify weak subjectivity checkpoint.
+		return errors.Wrap(err, "could not verify initial checkpoint provided for chain sync")
+	}
+
+	s.cfg.StateNotifier.StateFeed().Send(&feed.Event{
+		Type: statefeed.Initialized,
+		Data: &statefeed.InitializedData{
+			StartTime:             s.genesisTime,
+			GenesisValidatorsRoot: saved.GenesisValidatorRoot(),
+		},
+	})
+
+	s.spawnProcessAttestationsRoutine(s.cfg.StateNotifier.StateFeed())
+
+	return nil
+}
+
+func (s *Service) originRootFromSavedState(ctx context.Context) ([32]byte, error) {
+	// first check if we have started from checkpoint sync and have a root
+	originRoot, err := s.cfg.BeaconDB.OriginBlockRoot(ctx)
+	if err == nil {
+		return originRoot, nil
+	}
+	// if it's an ErrNotFound, that probably just means the node was synced from genesis
+	if !errors.Is(err, db.ErrNotFound) {
+		return originRoot, errors.Wrap(err, "could not retrieve checkpoint sync chain origin data from db")
+	}
+
+	// if the chain started from a genesis state, we should have a value for GenesisBlock
+	genesisBlock, err := s.cfg.BeaconDB.GenesisBlock(ctx)
+	if err != nil {
+		return originRoot, errors.Wrap(err, "could not get genesis block from db")
+	}
+	if err := helpers.BeaconBlockIsNil(genesisBlock); err != nil {
+		return originRoot, err
+	}
+	genesisBlkRoot, err := genesisBlock.Block().HashTreeRoot()
+	if err != nil {
+		return genesisBlkRoot, errors.Wrap(err, "could not get signing root of genesis block")
+	}
+	return genesisBlkRoot, nil
+}
+
+// initializeHeadFromDB uses the finalized checkpoint and head block found in the database to set the current head
+// note that this may block until stategen replays blocks between the finalized and head blocks
+// if the head sync flag was specified and the gap between the finalized and head blocks is at least 128 epochs long
+func (s *Service) initializeHeadFromDB(ctx context.Context) error {
+	finalized, err := s.cfg.BeaconDB.FinalizedCheckpoint(ctx)
+	if err != nil {
+		return errors.Wrap(err, "could not get finalized checkpoint from db")
+	}
+	if finalized == nil {
+		// This should never happen. At chain start, the finalized checkpoint
+		// would be the genesis state and block.
+		return errors.New("no finalized epoch in the database")
+	}
+	finalizedRoot := s.ensureRootNotZeros(bytesutil.ToBytes32(finalized.Root))
+	var finalizedState state.BeaconState
 
-		// Resume fork choice.
-		s.justifiedCheckpt = ethpb.CopyCheckpoint(justifiedCheckpoint)
-		s.prevJustifiedCheckpt = ethpb.CopyCheckpoint(justifiedCheckpoint)
-		s.bestJustifiedCheckpt = ethpb.CopyCheckpoint(justifiedCheckpoint)
-		s.finalizedCheckpt = ethpb.CopyCheckpoint(finalizedCheckpoint)
-		s.prevFinalizedCheckpt = ethpb.CopyCheckpoint(finalizedCheckpoint)
-		s.resumeForkChoice(justifiedCheckpoint, finalizedCheckpoint)
+	finalizedState, err = s.cfg.StateGen.Resume(ctx, s.cfg.FinalizedStateAtStartUp)
```

Comment:
There is a bit of an edge case here:
https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/state/stategen/service.go#L88
https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/state/stategen/service.go#L98

If the root in the db is a zero hash, would we be fine with returning this function returning a genesis state ? The reason I am asking, is because in the whole service, genesis root has been replaced with origin root. I am assuming the same logic also applies to stategen. 

Or are we assuming here, that in the case of `WSS`, the checkpoints are already saved into the db, and this edge case isnt really valid/possible ?

Reply:
> are we assuming here, that in the case of WSS, the checkpoints are already saved into the db, and this edge case isnt really valid/possible
yes, good point, I am making that assumption wrt stategen. I almost undertook a bigger refactor to split the block chain service into separate genesis/checkpoint types, but my instinct was that this zero hash logic is so prevalent and subtle that I didn't want to make huge and confusing change all at once. I also had a branch where I implemented an "origin store" that would get either the checkpoint sync block or genesis, but it started growing in scope so I fell back to this simpler option.

This change came in while I was working on the branch and is a little out of step with the concepts in this PR:
https://github.com/prysmaticlabs/prysm/blob/790bf031234f5bbd734fdb23f892e1041d293e4c/beacon-chain/node/node.go#L453
As an extra defensive check, in the code path linked above, where we detect the zero hash and fall back to genesis, we could first try to look up the origin block root via `OriginBlockRoot`, and fail if one exists (because that indicates that somehow we tried to start up with an origin checkpoint, but something went wrong with the finalization db entry). If we feel the need to do this check in multiple places I might try out the origin store idea again.

Reply:
Makes sense, thanks for the explanation ! I think having the check at only one place is fine now, as the assumption is always that the new weak subjectivity checkpoints are baked in.